<launch>

    <!-- SYSTEM MODEL -->
    <!-- vehicle parameters -->
    <param name="L_a" type="double" value="0.125" />
    <param name="L_b" type="double" value="0.125" />
    <param name="m" type="double" value="1.98" />
    <param name="I_z" type="double" value="0.24" />
    <param name="dt" type="double" value="0.1" />
    <param name="num_agents" type="int" value="1" />

    <!-- MACHINES -->
    <machine name="yoga" address="127.0.0.1" user="lukas" timeout="20.0" env-loader="/home/lukas/barc/env_loader_yoga.sh" />
    <!-- <machine name="yoga" address="192.168.10.74" user="lukas" timeout="20.0" env-loader="/home/lukas/barc/env_loader_yoga.sh" /> -->
    <machine name="ubuntu-pc" address="192.168.10.147" user="mpcubuntu" timeout="20.0" env-loader="/home/mpcubuntu/barc/env_loader_pc_vis.sh"/>
    <machine name="odroid_2" address="192.168.10.101" user="odroid" timeout="20.0" env-loader="/home/odroid/barc/env_loader_odroid.sh"/>

    <!-- AGENT 2 NODES-->
    <!-- AGENT-NODE 2-->
    <node machine="ubuntu-pc" pkg="barc" type="agent_node.jl" name="agent_2" output="screen">
        <param name="index" type="int" value="2" />
        <param name="color" type="str" value="blue" />
        <param name="mass" type="double" value="1.98" />
        <param name="time_offset" type="double" value="0.0" />
        <param name="delay_a" type="int" value="1" />
        <param name="delay_df" type="int" value="1" />
        <param name="sim" type="str" value="estimation" />

        <remap from="pos_info" to="agent_2/pos_info" />
        <remap from="real_val" to="agent_2/real_val" />
        <remap from="ecu" to="agent_2/ecu" />
        <remap from="xy_prediction" to="agent_2/xy_prediction" />
        <remap from="selected_states" to="agent_2/selected_states" />
        <remap from="prediction" to="agent_2/prediction" />
        <remap from="adv_prediction" to="agent_1/prediction_for_adv" />
        <remap from="prediction_for_adv" to="agent_2/prediction_for_adv" />
        <remap from="ecu_pwm" to="agent_2/ecu_pwm" />
    </node>

    <!-- Estimator -->
    <node machine="ubuntu-pc" pkg="barc" type="state_estimator_lukas_yaw.py" name="state_estimator_2" output="screen">
        <!-- delay in state estuiamtor -->
        <param name="delay_df" type="double" value="0.0" />
        <param name="delay_a" type="double" value="0.0" />
        <!-- ekf properties -->
        <param name="Q_x" type="double" value="0.01" />
        <param name="Q_y" type="double" value="0.01" />
        <param name="Q_vx" type="double" value="0.01" />
        <param name="Q_vy" type="double" value="0.01" />
        <param name="Q_ax" type="double" value="0.03" />
        <param name="Q_ay" type="double" value="0.05" />
        <param name="Q_psi" type="double" value="0.01" />
        <param name="Q_psiDot" type="double" value="0.01" />
        <param name="R_x" type="double" value="1.0" />
        <param name="R_y" type="double" value="1.0" />
        <param name="R_vx" type="double" value="1.0" />
        <param name="R_vy" type="double" value="10.0" />
        <param name="R_ax" type="double" value="2.0" />
        <param name="R_ay" type="double" value="8.0" />
        <param name="R_psiDot" type="double" value="0.5" />

        <remap from="hedge_pos" to="agent_2/hedge_pos" />
        <remap from="imu/data" to="agent_2/imu/data" />
        <remap from="vel_est" to="agent_2/vel_est" />
        <remap from="pos_info" to="agent_2/pos_info" />
        <remap from="ecu" to="agent_2/ecu" />
        <remap from="hedge_pos_ang" to="agent_2/hedge_pos_ang" />
        <!--<remap from="hedge_imu_fusion_1" to="agent_2/hedge_imu_fusion_1" /> -->
        <!--<remap from="hedge_imu_fusion_2" to="agent_2/hedge_imu_fusion_2" /> -->
        <remap from="hedge_imu_fusion" to="agent_2/hedge_imu_fusion" />
        <!-- ukf properties -->
    </node>


    <!-- IMU -->
    <node machine="odroid_2" pkg="myahrs_driver" type="myahrs_driver" name="myahrs_driver_2">
        <param name="port" value="/dev/ttyACM0" />
        <param name="baud_rate" value="115200" />

        <remap from="imu/data" to="agent_2/imu/data" />
        <remap from="imu/data_raw" to="agent_2/imu/data_raw" />
        <remap from="imu/mag" to="agent_2/imu/mag" />
        <remap from="imu/temperature" to="agent_2/imu/temperature" />
        <remap from="tf" to="agent_2/tf" />
    </node>

    <!-- GPS -->
    <node machine="odroid_2" pkg="marvelmind_nav" type="hedge_rcv_bin" name="indoor_gps_1" args="/dev/ttyACM1" >
        <remap from="beacons_pos_a" to="agent_2/beacons_pos_a" />
        <remap from="hedge_pos" to="agent_2/hedge_pos" />
        <remap from="hedge_pos_a" to="agent_2/hedge_pos_a" />
        <remap from="hedge_pos_ang" to="agent_2/hedge_pos_ang" />
        <!-- <remap from="hedge_imu_fusion_1" to="agent_2/hedge_imu_fusion_1" /> -->
        <remap from="hedge_imu_fusion" to="agent_2/hedge_imu_fusion" />
    </node>

    <!-- ARDUINO -->
    <!--  * encoders and ultrasound sensors -->
    <node machine="odroid_2" pkg="rosserial_python" type="serial_node.py" name="arduino_node_2" >
        <param name="port" value="/dev/ttyUSB0" />
        <remap from="vel_est" to="agent_2/vel_est" />
        <remap from="ecu_pwm" to="agent_2/ecu_pwm" />
        <remap from="diagnostics" to="agent_2/diagnostics" />
    </node>
    <!-- AGENT 2 NODES-->

    <!-- VISUALIZATION -->
    <node machine="ubuntu-pc" pkg="barc" type="plotter.py" name="visualization" output="screen">
        <param name="num_agents" type="int" value="1" />
        <param name="index" type="int" value="2" />
        <param name="agent_2" type="str" value="estimation" />
    </node>

</launch>
